internal/controllers/machine/machine_controller.go

```diff
@@ -252,6 +255,14 @@ func (r *Reconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster,
 		})
 	}
 
+	ok, err := r.isVersionAllowed(ctx, m, cluster)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to check machine version")
+	}
+	if !ok {
+		return ctrl.Result{}, errors.Wrap(err, "machine version is not allowed")
+	}
```

Comment:
I don't think this is right, wouldn't this code block reconciliation of a Machine as soon as the control plane bumps the version?

Reply:
In general, we should have checks in a validation webhook, potentially this one can be carrying a client that looks at the control plane reference

Reply:
Webhook on which resource? MachineDeployment, MachinePool, MachineSet, Machine or all of them?

Reply:
ha I just commented something similar here: https://github.com/kubernetes-sigs/cluster-api/issues/6040#issuecomment-1071824427

IMO the validation should be in the webhook of every object that can't change before the ControlPlane does but that has a Version in spec that can be edited by the user: Machine, MachinePool, MachineDeployment, and MachineSet.

I was thinking we'd want to validate updates too. If a user tries to upgrade a MachinePool or MachineDeployment before upgrading the ControlPlane that's breaking. What's your concern with `blocking reconciliation of existing machines because a control plane upgrade is in progress` @sbueringer? Wouldn't the new version be reflected in the control plane spec if the control plane has already been updated? Also, we wouldn't block reconciliation, just updating the Machine/MD/MP/MS version.

Reply:
For Machines, we need to make sure to not block Machine objects that are being created as part of the reconciliation of machine-based control planes, otherwise the control plane is never going to be rolled out

Reply:
good point, I think we can exclude control plane machines from this validation based on the control plane label?

Reply:
sgtm!